### PR TITLE
Add built-in package management REST API

### PR DESCRIPTION
**Pull request by joewiz**
## Summary

Add a built-in package management REST API to eXist-db, eliminating the need for the `existdb-packageservice` XAR package. The servlet handles package CRUD operations — list, get, install (from registry or upload), remove (with dependency checking), update-check, and icon serving.

## What Changed

All new code is in the **`exist-services`** Maven module (not `exist-core`), per the architectural principle that exist-core should contain only the core database engine.

### New module: `exist-services/`
- `PackageManagementServlet.java` — HTTP servlet at `/api/packages` for package CRUD
- `PackageService.java` — business logic for package operations (install, remove, dependency resolution)
- `RepoPackageLoader.java` — downloads packages from public/private registries

### Configuration
- `web.xml` — servlet registration with multipart config for XAR uploads
- `controller-config.xml` — URL pattern forwarding for `/api/packages`
- `pom.xml` — new `exist-services` module in reactor and `exist-distribution` dependency

## Motivation

The `existdb-packageservice` XAR package has a fundamental problem: it can't upgrade itself. The XQuery code executing the upgrade gets deleted mid-operation when the package is removed. By moving package management into the platform (compiled Java), this self-upgrade problem is eliminated.

## Test Plan

- [ ] `mvn install -pl exist-services -am -DskipTests` compiles cleanly
- [ ] Dashboard package management UI works (Cypress: `packages.cy.js`)
- [ ] `xst package install/remove/list` works against the built-in API
- [ ] public-repo app works (Cypress: all specs pass)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by adamretter** [comment]
May I suggest that this should go into its own separate module. The exist-core module was created with the specific purpose of being slimmed down rather than made bigger, the previous plans was to remove everything from exist-core apart from core database engine.

**Comment by line-o** [comment]
@joewiz I am intrigued by the approach you chose here. I assume that this does integrate with your other proposals to extend and unify the REST API exposed by exist-db.

I do see testfailures.

Since xst is not using the packagemanager/exist-packageservice at all it can be removed from the testplan.
This is not implying at all that xst could not benefit from a good API around packages. But ultimately this would need to be implemented in node-exist and xst and gulp-exist would both improve.

**Comment by joewiz** [comment]
[This response was co-authored with Claude Code. -Joe]

@line-o — test failures fixed. Three root causes:

1. **License check**: `exist-services/pom.xml` was missing the LGPL header
2. **Unit test failures** (URLRewritingTest, ControllerTest, JmxRemoteTest all 503): `PackageManagementServlet` was registered in `web.xml` with `load-on-startup`, causing Jetty to fail when the servlet class wasn't on exist-core's test classpath. Fix: moved servlet registration from `web.xml` to a `web-fragment.xml` in the exist-services JAR — only loaded when the JAR is present
3. **Integration test** (dependency:analyze): exist-services used 4 transitive deps without declaring them. Now declared explicitly.

Also made `Forward.java` return 404 gracefully instead of throwing when a target servlet is missing (defensive improvement).

6,559 tests, 0 failures.

Re: xst — agreed, the test plan reference was misleading. Removed.

Re: @adamretter's module suggestion — that's a good point about keeping exist-core slim. The exist-services module in this PR is now separate from exist-core.

**Comment by line-o** [comment]
@joewiz should this be added to existdb-openapi ?

**Comment by duncdrum** [comment]
my main concern is that since we will install dashboard on way or the other, we ll have two package services running? Should this be part of a larger group of PRs?

File: exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+                        gen.writeString(String.valueOf(item));
+                    }
+                }
+                gen.writeEndArray();
+            } else if (value instanceof Map<?, ?> nested) {
+                gen.writeFieldName(key);
+                writeMapAsJson(gen, (Map<String, Object>) nested);
+            } else {
+                gen.writeStringField(key, String.valueOf(value));
+            }
```
Comment (reinhapa) [inline exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java]: This could be changed to a switch expression

File: exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+        }
+        gen.writeEndObject();
+    }
+
+    private String normalizePath(final String pathInfo) {
+        if (pathInfo == null) {
+            return null;
+        }
+        // URL-decode the path
+        return java.net.URLDecoder.decode(pathInfo, java.nio.charset.StandardCharsets.UTF_8);
```
Comment (reinhapa) [inline exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java]: Use imports instead of canonical class names

File: exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+            writeError(response, HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
+                    "PACKAGE_ERROR", e.getMessage());
+        } catch (final EXistException e) {
+            writeError(response, HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
+                    "DATABASE_ERROR", e.getMessage());
+        }
+    }
+
+    @Override
+    protected void doDelete(final HttpServletRequest request, final HttpServletResponse response)
```
Comment (reinhapa) [inline exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java:140]: Address Codacy issue: _The method 'doDelete(HttpServletRequest, HttpServletResponse)' has an NPath complexity of 216, current threshold is 200_

File: exist-services/src/main/java/org/exist/repo/PackageService.java [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+
+/**
+ * Business logic for package management operations.
+ * Used by {@link org.exist.http.servlets.PackageManagementServlet} to provide
+ * a built-in REST API for package CRUD that does not depend on any XAR package.
+ */
+public class PackageService {
+
+    private static final Logger LOG = LogManager.getLogger(PackageService.class);
+
```
Comment (reinhapa) [inline exist-services/src/main/java/org/exist/repo/PackageService.java:58]: remove this empty line

File: exist-services/src/main/java/org/exist/repo/RepoPackageLoader.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+ * Loads packages from a remote public repository (e.g. https://exist-db.org/exist/apps/public-repo).
+ * Used by {@link Deployment} to resolve package dependencies during installation,
+ * and by {@link PackageService} for direct package installation from a registry.
+ *
+ * @param repoURL the base URL of the public package repository
+ */
+public record RepoPackageLoader(String repoURL) implements PackageLoader {
+
+    private static final Logger LOG = LogManager.getLogger(RepoPackageLoader.class);
+
```
Comment (reinhapa) [inline exist-services/src/main/java/org/exist/repo/RepoPackageLoader.java:50]: remove empty line

**Review by reinhapa — CHANGES_REQUESTED** [review]

**Comment by joewiz** [comment]
[This text was co-authored with Claude Code. -Joe]

This comment was prompted by @line-o and @duncdrum's questions above:

> should this be added to existdb-openapi ?

> since we will install dashboard on way or the other, we ll have two package services running?

## Package management: why it needs a home in core, and how it converges with existdb-openapi

**The motivation first — why not just leave package management to a package?** Because a package can't safely manage itself. existdb-openapi's `modules/packages.xqm` runs `repo:remove` / `repo:install-and-deploy` from XQuery stored *in the existdb-openapi XAR*. The moment you use that handler to upgrade or remove existdb-openapi itself (or anything it depends on), it deletes its own executing code mid-call — `remove` even clears the XQuery pool. So there must be at least one package-management entry point that is **not itself a package**: compiled, in core, able to manage any package including the package layer itself. That bootstrap case is the irreducible reason this PR exists.

**So why two surfaces, and aren't they redundant?** They're complementary, divided by *self-management capability* rather than by feature:

- **Core `/api/packages` (this PR)** — a compiled-Java `PackageService` behind `PackageManagementServlet` (new `exist-services` module), at server-root. The bootstrap-safe platform floor: always present, survives any package being removed, and the only surface that can safely upgrade/remove existdb-openapi itself.
- **existdb-openapi `/api/packages` (XQuery/Roaster, + existdb-openapi#49)** — the app-level surface for the common case (installing/removing *other* packages), consistent with the rest of the openapi API that clients (Oxygen plugin, eXide, vscode-existdb, the notebook kernel) already speak.

The operation sets are 1:1 (list / get / icon / install-from-registry / install-from-upload / remove / update-check) and there's no path collision (server-root vs app-scoped), so nothing is blocked today — the question is how they should relate.

### Direction: one engine in core; no duplicate logic

The thing to avoid is *two implementations* of package management drifting apart. So core owns the engine and the bootstrap-safe entry point — that part is non-negotiable for the self-upgrade case. What's genuinely open (see below) is the **end state of the app surface**: after convergence, should existdb-openapi drop its package endpoints in favor of the single core surface, or keep them as a **thin delegator** to the core engine for API-surface uniformity? Either way there is exactly one implementation. Our lean is the single core engine as the canonical implementation; whether the app keeps a delegating front door is the softer, uniformity-driven call.

This isn't a stub standing in for a real implementation: #6247 already authenticates and enforces `hasDbaRole()` (403) on install and remove, and `installFromUpload` already parses `multipart/`. So this is convergence/migration, not new capability.

### The work: three gates + sequencing

1. **Contract alignment.** The two surfaces differ in details that would silently break a consumer re-pointing between them — clearest example, the multipart field name: existdb-openapi#49 takes the file as `file` (Roaster's form-data shape), while this servlet reads `request.getPart("xar")`. These, plus the JSON shapes for list/get/icon and the error envelope, need reconciling so `/api/packages` is a drop-in for what the openapi consumers expect.
2. **OpenAPI description of `/api/packages`** so consumers keep the documented, generated-client-friendly contract they had under Roaster (see core-input item 2).
3. **Consumer migration.** Inventory and re-point everything that calls the openapi package endpoints (or `repo:*` directly): the Oxygen plugin, eXide, vscode-existdb, the notebook kernel, and the `existdb-packageservice` / Dashboard app.

**Sequence — do not retire existdb-openapi's package endpoints (or close existdb-openapi#49) first.** Land #6247 → migrate consumers to `/api/packages` → *then* settle the app-surface end state and retire existdb-openapi#49's standalone XQuery install path in favor of the core engine. Closing #49 before `/api/packages` is consumer-ready would open a regression window on the Oxygen plugin's "Install `.xar`…" feature. If that feature is wanted *in* the next beta and #6247 can't be merged-and-migrated in time, existdb-openapi#49 can ship as a documented interim and the openapi endpoints be removed (or converted to delegation) as a fast-follow.

### Needs core-team input (not settled between the strategy threads)

1. **App-surface end state — retire or delegate?** After convergence, should existdb-openapi drop its package endpoints (single core surface) or keep them as a thin delegator to the core engine? This is the "why two?" decision; the bootstrap mechanism in core is required either way, but the app-surface choice is a product/architecture call. (If we want delegation, that's the optional follow-up of exposing `PackageService` as an XQuery function module — useful for in-process callers that are *not* self-managing, but never the self-management path.)
2. **Should `/api/packages` be served/described via the built-in OpenAPI-routing direction (#6248), and is that direction wanted in core?** Gate 2 quietly reframes this PR from "a servlet" into "the first consumer of a built-in OpenAPI-routing layer in eXist core" — an architecture-of-core decision the team should buy into. Refinement we'd suggest: keep the requirement as "the endpoint must be OpenAPI-described," but **don't block #6247 on the full #6248 framework** — an interim hand-written description is acceptable if #6248 lags, so package-management convergence isn't held hostage to a larger piece.
3. **Security review of the new root servlet.** A server-root `/api/packages` that ships CORS headers and performs dba-gated install/remove is a security-sensitive core addition and wants a core security reviewer's eyes — CORS scope, CSRF for any cookie-auth path, and dba enforcement on every mutating op (and ideally the read ops too).

Feedback on the direction, and especially the three items above, would be much appreciated before we wire up the contract alignment and consumer migration.

**Comment by joewiz** [comment]
[This response was co-authored with Claude Code. -Joe]

Thanks @reinhapa — all addressed in 2f38194a1e:

- **Switch expression** — `writeMapAsJson` now uses pattern-matching `switch` instead of the `instanceof` if/else chains (outer dispatch and the nested list-item loop).
- **Canonical class names** — added imports for `URLDecoder` and `StandardCharsets`; the inline `java.net.`/`java.nio.charset.` references are gone.
- **`doDelete` NPath (216)** — now under the threshold, by extracting `writeDependentsConflict()` and `writePackageError()`.
- **Empty lines** — removed the blank between the type declaration and the `LOG` field in both `PackageService` and `RepoPackageLoader`.

While in `PackageService` I also brought `readExpathMetadata` under the NPath threshold (it was 262 — same file, not separately flagged) by extracting `readDependency()` / `putAttributeIfPresent()`, so Codacy stays clean across the changed files. `exist-services` compiles and Codacy PMD is clean on all four changed files.

Separately: this branch is now well behind `develop`. I've kept it on its original base for now so the review diff stays focused, but I'm happy to rebase onto current `develop` whenever you'd like it — just say the word (or if you'd rather review first and rebase right before merge, that works too).

**Comment by duncdrum** [comment]
@joewiz needs a rebase

**Comment by joewiz** [comment]
@duncdrum Ok, rebased.

**Comment by duncdrum** [comment]
Thank you @joewiz two failures on CI.

**Comment by joewiz** [comment]
[This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

Thanks @duncdrum — both were real, and both are fixed in 70f1d4a2a9.

**Native XSD checks.** The rebase brought the branch under the schema governance gate, and adding the `/api/packages` forward to the canonical `controller-config.xml` counts as a template change: `Template ... changed but @version is still "1.1.1"`. Bumped `controller-config.xsd` to 1.1.2 and kept the `schemaVersion` attribute on both canonical instances in step. I made it a PATCH rather than a MINOR because only the template's content changed — `forward` was already allowed by the grammar, so no new element or attribute is introduced and existing instance documents stay valid. Shout if you'd rather see MINOR.

**Windows integration.** This one turned out to be a latent bug in the branch rather than rebase fallout: `exist-services/pom.xml` pinned its parent to the literal `7.0.0-SNAPSHOT` instead of `${revision}` like every other module, so Maven failed while scanning projects with `Non-resolvable parent POM ... 'parent.relativePath' points at wrong local POM`. Linux never caught it because `exist-parent` was already in the local repository; building `exist-services` against an empty local repo reproduces it exactly, and switching to `${revision}` clears it.

Verified locally before pushing: `prepare-governance-context.sh` + `mvn -N xml:transform@schema-governance` against current `develop` passes, and `mvn validate` is clean across the reactor including the canonical-instance validation against the bumped schema.

One other thing worth flagging from the rebase, since it isn't visible in the diff: `develop` now generates each module's test `controller-config.xml` from the canonical template via `controller-config-fixture.xsl`, so this branch's old hand-edits to `exist-core/src/test/resources/standalone-webapp/WEB-INF/controller-config.xml` are gone rather than re-applied. All six fixtures use explicit `keep-forwards` allow-lists, so `/api/packages` is filtered out of the generated test configs and there's no test-side impact.

**Comment by duncdrum** [comment]
Thank you @joewiz. About those handrolled edits, if we don't need them in the test suite, leaving them out makes sense. If we should add them back in, the way is via xsl.

**Review by duncdrum — COMMENTED** [review]
Nice work standing up the built-in package API! Left a handful of inline notes — mostly a few unguarded edge cases that can throw uncaught exceptions, two temp-file leaks, a missing-tests flag, and some design/duplication points worth a look.

File: exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+        gen.writeEndObject();
+    }
+
+    private String normalizePath(final String pathInfo) {
+        if (pathInfo == null) {
+            return null;
+        }
+        // URL-decode the path
+        return URLDecoder.decode(pathInfo, StandardCharsets.UTF_8);
+    }
```
Comment (duncdrum) [inline exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java:441]: `request.getPathInfo()` is already URL-decoded by the servlet container, so decoding it again here double-decodes valid paths (a literal `+` becomes a space) and throws an uncaught `IllegalArgumentException` for a bare `%` not followed by two hex digits — e.g. `GET /api/packages/my%pkg` would 500 instead of returning a clean 4xx, since only `PackageException` is caught in the callers.

```suggestion
    private String normalizePath(final String pathInfo) {
        // getPathInfo() is already URL-decoded by the servlet container — decoding
        // it again double-decodes valid paths and throws on a bare '%'.
        return pathInfo;
    }
```

If you go this route, `java.net.URLDecoder` becomes an unused import and can be dropped too.

File: exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+        final String pathInfo = normalizePath(request.getPathInfo());
+
+        try {
+            if (pathInfo == null || pathInfo.isEmpty() || pathInfo.equals("/")) {
+                // GET /api/packages — list all
+                handleListPackages(response);
+            } else if (pathInfo.endsWith("/icon")) {
+                // GET /api/packages/{name}/icon
+                final String name = pathInfo.substring(1, pathInfo.length() - "/icon".length());
+                handleGetIcon(response, name);
```
Comment (duncdrum) [inline exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java:94]: `GET /api/packages/icon` (no package name) makes `pathInfo` exactly `"/icon"`, so `substring(1, 5 - 5)` becomes `substring(1, 0)` and throws `StringIndexOutOfBoundsException` — uncaught, so it surfaces as a 500 instead of a 400.

```suggestion
            } else if (pathInfo.endsWith("/icon")) {
                // GET /api/packages/{name}/icon
                final String name = pathInfo.substring(1, pathInfo.length() - "/icon".length());
                if (name.isEmpty()) {
                    writeError(response, HttpServletResponse.SC_BAD_REQUEST,
                            "BAD_REQUEST", "Package name is required");
                } else {
                    handleGetIcon(response, name);
                }
```

File: exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+            }
+            try (final InputStream is = xarPart.getInputStream();
+                 final DBBroker broker = getPool().get(Optional.of(user));
+                 final Txn transaction = getPool().getTransactionManager().beginTransaction()) {
+                final Map<String, Object> result = packageService.installFromUpload(
+                        broker, transaction, is);
+                transaction.commit();
+                writeJsonMap(response, HttpServletResponse.SC_OK, result);
+            }
+        } catch (final ServletException e) {
```
Comment (duncdrum) [inline exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java:331]: Jetty/Tomcat throw `IllegalStateException` (not `ServletException`) from `getPart()` when an upload exceeds the configured `max-file-size`/`max-request-size`. That's not caught here or by `doPost`'s other catches, so a too-large XAR upload 500s instead of getting a clean 400.

```suggestion
        } catch (final ServletException | IllegalStateException e) {
```

File: exist-core/src/main/java/org/exist/http/urlrewrite/Forward.java [hunk 9]
```diff
@@ -44,7 +48,9 @@ protected Forward(final URLRewrite other) {
     public void doRewrite(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
         final RequestDispatcher dispatcher = getRequestDispatcher(request);
         if (dispatcher == null) {
-            throw new ServletException("Failed to initialize request dispatcher to forward request to " + uri);
+            LOG.warn("No servlet registered for forward target '{}' — is the required module on the classpath?", uri);
+            response.sendError(HttpServletResponse.SC_NOT_FOUND, "Servlet not available: " + uri);
+            return;
         }
```
Comment (duncdrum) [inline exist-core/src/main/java/org/exist/http/urlrewrite/Forward.java:54]: Heads up: `Forward` is the shared base class for *every* `<forward>` entry in `controller-config.xml`, not just the new package API route — so this change to log-and-404 instead of throwing softens failures app-wide. A typo'd servlet name in `/rest`, `/webdav`, etc. would previously surface as a loud 500/`ServletException` (visible to monitoring); now it silently degrades to a per-request 404 with only a WARN log line, which could make a broken deployment look like ordinary "not found" traffic.

If the goal is just to let the package API's forward target be optional (module not on the classpath), might be worth scoping the graceful-404 behavior to that specific subclass instead of changing the shared base class's contract for all forwards.

File: exist-services/src/main/java/org/exist/repo/PackageService.java [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+        final int startQuote = json.indexOf('"', colonIdx + 1);
+        if (startQuote < 0) {
+            return null;
+        }
+        final int endQuote = json.indexOf('"', startQuote + 1);
+        if (endQuote < 0) {
+            return null;
+        }
+        return json.substring(startQuote + 1, endQuote);
+    }
```
Comment (duncdrum) [inline exist-services/src/main/java/org/exist/repo/PackageService.java:473]: This hand-rolled substring scan has no awareness of JSON nesting or escaping, so a body like `{"other":{"name":"nested"},"url":"..."}` or a value containing an escaped quote can make it find the wrong/truncated occurrence of a key and silently return an incorrect value instead of failing loudly. `jackson-core` is already a dependency in this module (used for `JsonGenerator` in the servlet) — its streaming `JsonParser` would avoid this class of bug cheaply:

```suggestion
    public static String extractJsonStringValue(final String json, final String key) {
        try (final com.fasterxml.jackson.core.JsonParser parser =
                     new com.fasterxml.jackson.core.JsonFactory().createParser(json)) {
            if (parser.nextToken() != com.fasterxml.jackson.core.JsonToken.START_OBJECT) {
                return null;
            }
            while (parser.nextToken() == com.fasterxml.jackson.core.JsonToken.FIELD_NAME) {
                final String fieldName = parser.currentName();
                parser.nextToken();
                if (fieldName.equals(key) && parser.currentToken() == com.fasterxml.jackson.core.JsonToken.VALUE_STRING) {
                    return parser.getText();
                }
                parser.skipChildren();
            }
            return null;
        } catch (final java.io.IOException e) {
            return null;
        }
    }
```

(Used fully-qualified names to keep this a self-contained suggestion — worth switching to proper imports if you take it.)

File: exist-services/src/main/java/org/exist/repo/PackageService.java [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+        // Read package name from the XAR descriptor
+        final Optional<org.exist.dom.memtree.DocumentImpl> descriptor = deployment.getDescriptor(broker, xarSource);
+        final String name = descriptor.map(d -> d.getDocumentElement().getAttribute("name")).orElse("unknown");
+
+        final Map<String, Object> result = new LinkedHashMap<>();
+        result.put("status", "ok");
+        result.put("name", name);
+        target.ifPresent(t -> result.put("target", t));
+        return result;
+    }
```
Comment (duncdrum) [inline exist-services/src/main/java/org/exist/repo/PackageService.java:179]: `tempFile` is never released back to `TemporaryFileManager` — only `deleteOnExit()` protects it (registered internally by the manager), which the manager's own javadoc documents as unreliable. Each upload leaks up to the XAR's size on disk until restart.

```suggestion
    public Map<String, Object> installFromUpload(final DBBroker broker, final Txn transaction,
                                                  final InputStream xarStream) throws PackageException, IOException {
        // Save to temporary file
        final TemporaryFileManager tempManager = TemporaryFileManager.getInstance();
        final Path tempFile = tempManager.getTemporaryFile();
        try {
            Files.copy(xarStream, tempFile, StandardCopyOption.REPLACE_EXISTING);

            final XarFileSource xarSource = new XarFileSource(tempFile);
            final Deployment deployment = new Deployment();
            final Optional<String> target = deployment.installAndDeploy(broker, transaction, xarSource, null);

            // Read package name from the XAR descriptor
            final Optional<org.exist.dom.memtree.DocumentImpl> descriptor = deployment.getDescriptor(broker, xarSource);
            final String name = descriptor.map(d -> d.getDocumentElement().getAttribute("name")).orElse("unknown");

            final Map<String, Object> result = new LinkedHashMap<>();
            result.put("status", "ok");
            result.put("name", name);
            target.ifPresent(t -> result.put("target", t));
            return result;
        } finally {
            tempManager.returnTemporaryFile(tempFile);
        }
    }
```

File: exist-services/src/main/java/org/exist/repo/RepoPackageLoader.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        connection.connect();
+
+        try (final InputStream is = connection.getInputStream()) {
+            final TemporaryFileManager temporaryFileManager = TemporaryFileManager.getInstance();
+            final Path outFile = temporaryFileManager.getTemporaryFile();
+            Files.copy(is, outFile, StandardCopyOption.REPLACE_EXISTING);
+            return new XarFileSource(outFile);
+        } catch (final IOException e) {
+            throw new IOException("Failed to download package from " + pkgURL + ": " + e.getMessage(), e);
+        }
```
Comment (duncdrum) [inline exist-services/src/main/java/org/exist/repo/RepoPackageLoader.java:87]: Same temp-file leak as `installFromUpload` below — `outFile` is never returned to `TemporaryFileManager`, relying only on `deleteOnExit()`. Every registry install and every dependency resolved during install (`PackageService.installFromRegistry` / `Deployment`) leaves a downloaded XAR on disk indefinitely.

Unlike the upload case, the fix isn't fully local: `load()` hands the `XarFileSource` (wrapping this temp file) back to the caller, which may still be reading from it during `installAndDeploy`. Cleanest fix is probably for the caller (`PackageService.installFromRegistry`) to track and release the temp path once `deployment.installAndDeploy(...)` returns, mirroring the `try/finally` suggested for `installFromUpload`.

File: exist-services/src/main/java/org/exist/repo/RepoPackageLoader.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+import java.nio.file.StandardCopyOption;
+
+/**
+ * Loads packages from a remote public repository (e.g. https://exist-db.org/exist/apps/public-repo).
+ * Used by {@link Deployment} to resolve package dependencies during installation,
+ * and by {@link PackageService} for direct package installation from a registry.
+ *
+ * @param repoURL the base URL of the public package repository
+ */
+public record RepoPackageLoader(String repoURL) implements PackageLoader {
```
Comment (duncdrum) [inline exist-services/src/main/java/org/exist/repo/RepoPackageLoader.java:48]: This duplicates the existing private inner `RepoPackageLoader` record in `extensions/modules/expathrepo/.../Deploy.java` almost verbatim (same query-URL construction, same download-to-temp-file logic) rather than extracting a shared implementation. They've already drifted slightly — different `User-Agent` string and exception cause-chaining — so a future fix to one (e.g. the temp-file leak above, or the query parameter format) won't automatically propagate to the other. Worth considering extracting one shared implementation that both `Deploy` and this module use.

File: exist-services/src/main/java/org/exist/repo/PackageService.java [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+        if (maybeRepo.isEmpty()) {
+            return Collections.emptyList();
+        }
+        final Repository repo = maybeRepo.get().getParentRepo();
+        final String findUrl = registryUrl + "/find";
+        final String processorVersion = org.exist.SystemProperties.getInstance()
+                .getSystemProperty("product-version", "7.0.0");
+
+        final List<Map<String, Object>> updates = new ArrayList<>();
+        for (final Packages packages : repo.listPackages()) {
```
Comment (duncdrum) [inline exist-services/src/main/java/org/exist/repo/PackageService.java:268]: `checkUpdates()` queries the registry for every installed package sequentially (10s connect + 10s read timeout each), with no concurrency and no overall bound on total latency. An instance with many packages and a slow/unresponsive registry can block `POST /api/packages/update-check` for up to N × ~20s. Worth considering a small bounded `ExecutorService` (or `HttpClient`'s async API) to fan these out, especially since each iteration is independent.

File: exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+ * <ul>
+ *   <li>GET /api/packages — list all installed packages</li>
+ *   <li>GET /api/packages/{name} — get single package details</li>
+ *   <li>GET /api/packages/{name}/icon — get package icon</li>
+ *   <li>POST /api/packages/install — install from registry (JSON) or upload (multipart)</li>
+ *   <li>POST /api/packages/update-check — check for available updates</li>
+ *   <li>DELETE /api/packages/{name} — remove package</li>
+ * </ul>
+ */
+public class PackageManagementServlet extends AbstractExistHttpServlet {
```
Comment (duncdrum) [inline exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java:66]: Flagging here too (see review summary): this new servlet — and the ~6 endpoints it exposes with specific status codes and JSON bodies — has no test coverage anywhere in the diff. Per this repo's guidance, HTTP/transport-layer behavior like this should get a `RESTServiceTest`-style integration test so regressions in status codes or JSON shape get caught.

File: exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java (hunk 6 above)
Comment (joewiz) [reply exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java]: [This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

Thanks — the two failures are real, but the decode itself turns out to be needed. The servlet is reached through the controller's `<forward>`, and the forwarded request's path info comes from `XQueryURLRewrite.RequestWrapper`, which takes it from the raw `getRequestURI()`. So it arrives still percent-encoded. Package names are usually URIs (`/api/packages/http%3A%2F%2Fexist-db.org%2Fapps%2Fdashboard`), so dropping the decode would break lookups by name.

What fc228593b8 changes instead: only percent escapes are decoded, so `+` stays a literal plus, and a malformed escape such as a bare `%` now answers `400` instead of escaping as a 500.

File: exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java (hunk 7 above)
Comment (joewiz) [reply exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java]: [This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

Fixed as you suggested: an empty name answers `400`.

File: exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java (hunk 8 above)
Comment (joewiz) [reply exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java]: [This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

Fixed, with two adjustments. It answers `413 Payload Too Large`, which says more than a 400. And the `IllegalStateException` is caught around `getPart` only: a catch around the whole install would also turn an unrelated `IllegalStateException` from the installation into "upload too large".

File: exist-core/src/main/java/org/exist/http/urlrewrite/Forward.java (hunk 9 above)
Comment (joewiz) [reply exist-core/src/main/java/org/exist/http/urlrewrite/Forward.java]: [This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

Agreed, and it reached further than `controller-config.xml`: `PathForward` also handles `<exist:forward servlet="…">` from `controller.xq`, so app URL rewriting was affected too. Now a forward answers 404 for a missing servlet only when it is marked `optional="true"`. Every other forward fails loudly as before. The package API's forward carries the attribute, and `controller-config.xsd` declares it (that's what the 1.1.2 schema bump now covers). `PathForwardOptionalTest` covers both cases.

File: exist-services/src/main/java/org/exist/repo/PackageService.java (hunk 10 above)
Comment (joewiz) [reply exist-services/src/main/java/org/exist/repo/PackageService.java]: [This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

Replaced with Jackson's streaming parser, as you suggested, with imports and a shared `JsonFactory`. It reads top-level members only. `PackageServiceJsonTest` covers a nested object with the same key, escaped quotes, and missing, non-string and malformed input.

File: exist-services/src/main/java/org/exist/repo/PackageService.java (hunk 11 above)
Comment (joewiz) [reply exist-services/src/main/java/org/exist/repo/PackageService.java]: [This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

Fixed with a `try`/`finally` that returns the file to `TemporaryFileManager` once the installation has read it.

File: exist-core/src/main/java/org/exist/repo/RepoPackageLoader.java [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+        connection.connect();
+
+        try (final InputStream is = connection.getInputStream()) {
+            final TemporaryFileManager temporaryFileManager = TemporaryFileManager.getInstance();
+            final Path outFile = temporaryFileManager.getTemporaryFile();
+            Files.copy(is, outFile, StandardCopyOption.REPLACE_EXISTING);
+            return new XarFileSource(outFile);
+        } catch (final IOException e) {
+            throw new IOException("Failed to download package from " + pkgURL + ": " + e.getMessage(), e);
+        }
```
Comment (joewiz) [reply exist-core/src/main/java/org/exist/repo/RepoPackageLoader.java]: [This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

Fixed in the loader itself rather than in each caller: `RepoPackageLoader` is now `AutoCloseable`. It records every file it downloads, including dependencies resolved during `installAndDeploy`, and `close()` returns them all. Both callers use it in a try-with-resources that spans the installation, so the files stay available as long as they're read.

File: exist-services/src/main/java/org/exist/repo/RepoPackageLoader.java (hunk 13 above)
Comment (joewiz) [reply exist-services/src/main/java/org/exist/repo/RepoPackageLoader.java]: [This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

Done. There's now one `RepoPackageLoader`, in exist-core's `org.exist.repo` next to `Deployment` and `PackageLoader`, and `repo:install-and-deploy` uses it instead of its private copy. That brings `Deploy` the temp-file fix and the exception cause-chaining too. On the User-Agent: I kept "eXist-db Package Manager" over the old Firefox 3.5 string; the public repo answers both the same way.

File: exist-services/src/main/java/org/exist/repo/PackageService.java (hunk 14 above)
Comment (joewiz) [reply exist-services/src/main/java/org/exist/repo/PackageService.java]: [This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

Done: the lookups run four at a time on a small fixed pool, with a 60-second bound on the whole check. Lookups still running at the deadline are cancelled, and the check reports what it has.

File: exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java (hunk 15 above)
Comment (joewiz) [reply exist-services/src/main/java/org/exist/http/servlets/PackageManagementServlet.java]: [This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

Agreed, it needs HTTP-level tests. This round adds unit tests for the pieces I changed (`PathForwardOptionalTest` in exist-core, `PackageServiceJsonTest` as exist-services' first test). A `RESTServiceTest`-style test of the endpoints needs the Jetty test harness exist-core uses, set up in exist-services, which has had none so far. I'll add that in a follow-up commit on this PR, covering status codes and JSON shapes for each endpoint, including the 400/413 cases above.

**Review by duncdrum — APPROVED** [review]
